routing/bandwidth.go

```diff
@@ -89,30 +130,129 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
+
+			log.Debugf("ShortChannelID=%v: external traffic "+
+				"shaper is handling traffic: %v", cid,
+				shouldHandle)
+
+			// If this channel isn't handled by the external traffic
+			// shaper, we'll return early.
+			if !shouldHandle {
+				return fn.Ok(bandwidthResult{})
+			}
+
+			// Ask for a specific bandwidth to be used for the
+			// channel.
+			commitmentBlob := link.CommitmentCustomBlob()
+			auxBandwidth, err := ts.PaymentBandwidth(
+				firstHopBlob, commitmentBlob, linkBandwidth,
+			)
+			if err != nil {
+				return bandwidthErr(fmt.Errorf("failed to get "+
+					"bandwidth from external traffic "+
+					"shaper: %w", err))
+			}
+
+			log.Debugf("ShortChannelID=%v: external traffic "+
+				"shaper reported available bandwidth: %v", cid,
+				auxBandwidth)
+
+			// We don't know the actual HTLC amount that will be
+			// sent using the custom channel. But we'll still want
+			// to make sure we can add another HTLC, using the
+			// MayAddOutgoingHtlc method below. Passing 0 into that
+			// method will use the minimum HTLC value for the
+			// channel, which is okay to just check we don't exceed
+			// the max number of HTLCs on the channel. A proper
+			// balance check is done elsewhere.
+			return fn.Ok(bandwidthResult{
+				bandwidth:  fn.Some(auxBandwidth),
+				htlcAmount: fn.Some[lnwire.MilliSatoshi](0),
+			})
+		},
+	).Unpack()
+	if err != nil {
+		log.Errorf("ShortChannelID=%v: failed to get bandwidth from "+
+			"external traffic shaper: %v", cid, err)
+
+		return 0
+	}
+
+	htlcAmount := amount
+	result.htlcAmount.WhenSome(func(amount lnwire.MilliSatoshi) {
+		htlcAmount = amount
+	})
+
+	// If our link isn't currently in a state where it can add another
 	// outgoing htlc, treat the link as unusable.
-	if err := link.MayAddOutgoingHtlc(amount); err != nil {
-		log.Warnf("ShortChannelID=%v: cannot add outgoing htlc: %v",
-			cid, err)
+	if err := link.MayAddOutgoingHtlc(htlcAmount); err != nil {
+		log.Warnf("ShortChannelID=%v: cannot add outgoing "+
+			"htlc with amount %v: %v", cid, htlcAmount, err)
 		return 0
 	}
 
-	// Otherwise, we'll return the current best estimate for the available
-	// bandwidth for the link.
-	return link.Bandwidth()
+	// If the external traffic shaper determined the bandwidth, we'll return
+	// that value, even if it is zero (which would mean no bandwidth is
+	// available on that channel).
+	reportedBandwidth := linkBandwidth
+	result.bandwidth.WhenSome(func(bandwidth lnwire.MilliSatoshi) {
+		reportedBandwidth = bandwidth
+	})
+
+	return reportedBandwidth
 }
 
 // availableChanBandwidth returns the total available bandwidth for a channel
 // and a bool indicating whether the channel hint was found. If the channel is
 // unavailable, a zero amount is returned.
 func (b *bandwidthManager) availableChanBandwidth(channelID uint64,
-	amount lnwire.MilliSatoshi) (lnwire.MilliSatoshi, bool) {
+	amount lnwire.MilliSatoshi,
+	firstHopBlob fn.Option[tlv.Blob]) (lnwire.MilliSatoshi, bool) {
```

Comment:
I think we'll want some better terminology than `firstHopBlob` here. I've been trying to figure out what to call it but the way I think about it is that the blob represents some kind of opaque "intent", "color", or "properties" or something.

I don't mind that the type is that of `tlv.Blob` but as usual I am advocating naming variables and fields by their semantic purpose rather than their representational encoding.

So the question here is that, at it's most abstract conceptualization, what *IS* the `firstHobBlob` or *any* of the "blob" named fields in this PR?

Reply:
> what IS the firstHobBlob or any of the "blob" named fields in this PR?

When it's `tlv.Blob`, it's just an opauqe TLV blob. It might go into a wire message in which case we morph it into `lnwire.CustomRecords`. When it doesn't go into a wire message it's something passed to a higher layer for intepretation. 

In this case, `firstHopBlob` is used to inform how path finding should function. If specified here, then we may only want to use channels that can handle that blob (based on the `TrafficShaper`). 

In short, the blobs are a way for `lnd` to maintain minimal knowledge w.r.t what the higher level is doing, while still being able to enforce normal TLV invariants such as no duplicate keys or the keys are in order.

Reply:
Is it possible to specify different asset types in `lnwire.CustomRecords` for payment splitting? It feels like there is something in analogy to a payment and its multiple parts where the initial custom records represent the payment and the modified custom records represent the different htlcs or parts.

Reply:
I understand the opacity. I also understand we don't want to make the opaque transparent. But there is some abstraction that covers all the blobs. After all, "blob" is a description of *representation*, not a description of *purpose*

Reply:
Maybe something like `BandwidthTypeSelector` comes close? In the asset sense we select the asset type and the amount, but it is not clear when used in a more generic way with other custom channel implementations if that's still accurate.

Reply:
But that's the thing: `lnd` doesn't even know what the purpose of this thing is. It's completely unknown. So IMO "blob" is both a representation _and_ the superset of what we know about it in this context.
Also, we've used this terminology for the data we store everywhere now, IMO it would be more confusing to start changing it now for just this purpose.